packages/vscode-webui/src/features/chat/page.tsx

```diff
@@ -114,7 +131,97 @@ function Chat({ user, uid, prompt, files }: ChatProps) {
 
   useRestoreTaskModel(task, isModelsLoading, updateSelectedModelId);
 
-  const { sendNotification } = useSendTaskNotification();
+  const { sendNotification, clearNotification } = useSendTaskNotification();
+
+  const { toolset } = useMcp();
+
+  const { autoApproveActive, autoApproveSettings } = useAutoApprove({
+    autoApproveGuard: autoApproveGuard.current === "auto",
+    isSubTask,
+  });
+
+  const { retryCount } = useRetryCount();
+
+  const onStreamFinish = useLatest(
+    (
+      data: Pick<Task, "id" | "cwd" | "status"> & {
+        messages: Message[];
+      },
+    ) => {
+      const lastMessage = data.messages.at(-1);
+      const taskUid = isSubTask ? task?.parentId : uid;
+      if (!taskUid || !lastMessage) return;
+
+      if (data.status === "pending-tool") {
+        const pendingToolCallApproval = getPendingToolcallApproval(lastMessage);
+        if (pendingToolCallApproval) {
+          const autoApproved = isToolAutoApproved({
+            autoApproveActive,
+            autoApproveSettings,
+            toolset,
+            pendingApproval: pendingToolCallApproval,
+          });
+
+          if (!autoApproved) {
+            sendNotification("pending-tool", { uid: taskUid, cwd: data.cwd });
+          }
+        }
+      }
+
+      if (data.status === "pending-input") {
+        const readyForRetryError = getReadyForRetryError(messages);
+        if (!readyForRetryError) return;
+
+        const retryLimit =
+          autoApproveActive && autoApproveSettings.retry
+            ? autoApproveSettings.maxRetryLimit
+            : 0;
+
+        if (
+          retryLimit === 0 ||
+          (retryCount?.count !== undefined && retryCount.count >= retryLimit)
+        ) {
+          sendNotification("pending-input", { uid: taskUid, cwd: data.cwd });
+        }
+      }
+
+      if (data.status === "completed") {
+        sendNotification("completed", { uid: taskUid, cwd: data.cwd });
+      }
+    },
+  );
+
+  const onStreamFailed = useLatest(
+    ({ error, cwd }: { error: Error; cwd: string | null }) => {
+      const taskUid = isSubTask ? task?.parentId : uid;
+      if (!taskUid) return;
+
+      let autoApprove = autoApproveGuard.current === "auto";
+      if (error && Object.values(PochiApiErrors).includes(error.message)) {
+        autoApprove = false;
+      }
+
+      if (
+        error &&
+        APICallError.isInstance(error) &&
+        error.isRetryable === false
+      ) {
```

Comment:
maybe we can have a `isRetryableError` util to reuse logic with https://github.com/TabbyML/pochi/blob/3919cd945b94e954e8e055020dde144dad5eae1a/packages/vscode-webui/src/features/approval/hooks/use-pending-retry-approval.ts#L59